R/input-time-series.R

```diff
@@ -0,0 +1,170 @@
+##' Prepare data for ART input time series plots
+##'
+##' Take uploaded ART and shape file paths and format as data which
+##' can be used to draw input time series graphs.
+##'
+##' @param art Path to file containing ART data
+##' @param shape Path to file containing geojson areas data
+##'
+##' @return Data formatted for plotting input time series containing columns
+##' area_id, area_name, area_level_label, time_step, time_period, plot and
+##' value
+##' @export
+prepare_input_time_series_art <- function(art, shape) {
+
+  ## From "area_id", "calendar_quarter", "sex", "age_group", "art_current"
+
+  ## To "area_id","area_name","area_level_label", "area_level", time_step",
+  ## "time_period","plot", "value"
+
+  ## Check if data is quarterly or annual - if quarterly aggregate to annual
+  areas <- sf::read_sf(shape) %>% sf::st_drop_geometry()
+
+  art_number <- readr::read_csv(art, show_col_types = FALSE) %>%
+    dplyr::left_join(areas %>% dplyr::select(area_id, area_level), by = "area_id") %>%
+    dplyr::mutate(year = year_labels(calendar_quarter_to_quarter_id(calendar_quarter)))
+
+  # Check for duplicated years in area_id/sex/age combinations
+  dup <- art_number %>%
+    dplyr::group_by(area_id, sex, age_group, year) %>%
+    dplyr::summarise(n=n(), .groups = 'drop')
+
+  if(unique(dup$n == 4)) {
+    # If quarterly values available for all disags - aggregate to annual
+    art_number_annual <- art_number %>%
+      dplyr::group_by(area_id, area_name, area_level, sex, age_group, year) %>%
+      dplyr::summarise(art_current = sum(art_current, na.rm = TRUE),
+                .groups = 'drop') %>%
+      dplyr::mutate(time_period = paste0("CY", year, "Q4"),
+             time_step = "annual")
+
+    # Bind to quarterly data set
+    art_number <- art_number %>%
+      dplyr::mutate(time_step = "quarterly") %>%
+      dplyr::rename(time_period = calendar_quarter) %>%
+      rbind(art_number_annual)
+  } else if (unique(dup$n == 1)) {
+    # If annual values available for all disags - don't aggregate
+    art_number <- art_number %>%
+      dplyr::mutate(time_step = "annual") %>%
+      dplyr::rename(time_period = calendar_quarter)
+  } else {
+    # Throw error for duplicate annual or incomplete quarterly data
+      stop(print("Quarterly data not provdied for all dissagregates/n
```

Comment:
@jeffeaton @r-ash Should we add a check in here that would throw an error and identify:
1. Missing quarterly data
2. Duplicate annual data
Or should user be able to upload incomplete data at this stage?

Reply:
So any data that gets here will have already had this function https://github.com/mrc-ide/hintr/blob/master/R/validate_inputs.R#L112 called on it. Or for anc https://github.com/mrc-ide/hintr/blob/master/R/validate_inputs.R#L149

So for ART
1. Not checking for missing quarterly data - we could add this as a check in the validate function
2. We're checking that there is a unique combination of `"area_id", "calendar_quarter", "sex", "age_group"` (not sure if that covers duplicate annual case?) 

For ANC we check that there are unique combinations of `"area_id", "age_group", "year"`.

If there are any data validation checks which we need to run then we should put them into the validate functions - so that users will get the error on upload instead of when trying to generate the graph.

Reply:
We should also use translated text for the error `stop(t_("KEY"))`